C#: Add tuple dataflow

tamasvajk opened the pull request:
This PR adds better data flow support for tuples. We're now tracking data flowing into tuple fields and out of them. Also, tuple extraction was slightly changed to remove named tuple fields.

~~[Differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/901/)~~ ~~[Differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/927/)~~ [Differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/937/)

tamasvajk commented on csharp/ql/test/library-tests/assignables/Assignables.expected:
```diff
@@ -39,39 +39,12 @@
 | Assignables.cs:78:22:78:22 | s |
 | Assignables.cs:82:21:82:24 | temp |
 | Assignables.cs:84:30:84:30 | e |
-| Assignables.cs:92:10:92:14 | Item1 |
```
These have been removed because they are not `fromSource()`. Does this cause any issue?

hvitved replied:
I think that is fine.

tamasvajk commented on csharp/ql/test/library-tests/dataflow/tuples/Tuples.cs:
```diff
@@ -0,0 +1,84 @@
+    static void M3()
+    {
+        var x = (ValueTuple<string, int, int>)("taint source", 2, 3);
+        Sink(x.Item1);          // Tainted
```
I'm a bit surprised that this is found.

hvitved replied:
Casts are modelled as normal value-preserving data-flow steps: https://github.com/github/codeql/blob/main/csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll#L188.

hvitved commented:
We should now be able to get rid of https://github.com/github/codeql/blob/main/csharp/ql/src/semmle/code/csharp/dataflow/internal/TaintTrackingPrivate.qll#L65-L71 and https://github.com/github/codeql/blob/main/csharp/ql/src/semmle/code/csharp/dataflow/internal/TaintTrackingPrivate.qll#L76-L82, right?

tamasvajk commented:
@hvitved I rebased this PR to fix the merge conflicts, fixed the nullability annotations in new code, and added an extra commit to handle patterns.

I started a new [differences job](https://jenkins.internal.semmle.com/job/Changes/job/CSharp-Differences/927/).

hvitved commented on csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:
```diff
@@ -828,6 +835,15 @@ private module Cached {
         )
+        or
+        // item = variable in node1 = (..., variable, ...) in a case/is var (..., ...)
+        exists(PatternExpr pe | pe.getAChild*() = te) and
```
I have a preference for `te = any(PatternExpr pe).getAChildExpr*()`

hvitved commented on csharp/ql/src/semmle/code/csharp/exprs/Expr.qll:
```diff
@@ -1040,7 +1040,7 @@ class TupleExpr extends Expr, @tuple_expr {
   predicate isConstruction() {
     not this = getAnAssignOrForeachChild() and
-    not this instanceof PatternExpr
+    not exists(PatternExpr pe | pe.getAChild*() = this)
```
Again I prefer `not this = any(PatternExpr pe).getAChildExpr*()`

hvitved reviewed (commented):
Excellent that we now handle tuple patterns as well.

tamasvajk commented:
@hvitved I added an extra commit with mixed positional and variable decl. tuple pattern. I'm not going to address that in this PR, but it shows what I'd like to cover next.
